File: aula7/banco.cpp
#include "banco.h"
#include <iostream>

using namespace std;

Banco::Banco() //O construtor criara 4 contas
{
    this->top = 3;
    this->contas = new Conta[top+1];
    this->senhaFuncionario = 246810;
    this->contas[0] = Conta(1234, 1, "Joao", "Corrente", 300);
    this->contas[1] = {4567, 2, "Jose", "Poupanca", 800};
    this->contas[2] = {7890, 3, "Maria", "Corrente", 1000};
    this->contas[3] = {8956, 4, "Madalena", "Poupanca", 2000};
}

Banco::~Banco()
{
    delete[] this->contas;
}

Conta *Banco::buscaConta(int numero)//Retorna o endereço da conta que possuir o mesmo numero informado
{
    for (int i = 0; i <= top; i++)
    {
        if (numero == this->contas[i].numero)
        {
            return &this->contas[i];
        }
    }

    return nullptr;
}

void Banco::atendimentoCliente() //Realiza o atendimento ao cliente(Função chamada na main)
{
    Conta *contaCliente;
    int numC = 0;
    int senhain;
    bool atendimento = true;

    cout << "Bem vindo ao sistema de atendimento do banco" << endl;
    cout << "Digite o numero da sua conta: ";
    cin >> numC;

    contaCliente = this->buscaConta(numC); //Chama o Metodo buscaConta() do banco para achar o objeto conta que possui o numero numC

    if (contaCliente == nullptr)//Se não achar nenhuma conta que corresponda entra nesse if
    {
        cout << "Conta invalida" << endl;
    }
    else
    {
        cout << "Digite a sua senha: ";
        cin >> senhain;

        if (contaCliente->validaSenha(senhain))
        {
            cout << "Ola " << contaCliente->titular << endl;
            while (atendimento) //Realiza o atendimento
            {
                int op;
                double valor;
                cout << "Qual operacao deseja fazer? (1 - Saque, 2 - Deposito, 3 - Ver Saldo, 4 - Transferencia, 5 - Sair): ";
                cin >> op;
                switch (op)
                {
                case 1:
                    cout << "Digite o valor: ";
                    cin>>valor;
                    contaCliente->saque(senhain,valor);
                    break;
                case 2:
                    cout << "Digite o valor: ";
                    cin>>valor;
                    contaCliente->deposito(valor);
                    break;
                case 3:
                    cout << "Saldo: R$ "<<contaCliente->getSaldo(senhain)<<endl;
                    break;
                case 4:
                    cout << "Digite o valor: ";
                    cin>>valor;
                    int numD;
                    cout << "Digite o número da conta de destino: ";
                    cin >> numD;
                    Conta *cdest;
                    cdest = this->buscaConta(numD);
                    contaCliente->transferencia(valor, senhain, cdest);
                    break;
                case 5:
                    atendimento = false;
                    break;
                }
            }
        }
        else
        {
            cout << "Senha invalida" << endl;
        }
    }
}

void Banco::atendimentoFuncionario(){
    int senhain;
    bool atendimento = true;
    cout << "Bem vindo ao sistema de atendimento do banco ao funcionário" << endl;
    cout << "Digite a sua senha: ";
    cin >> senhain;
    if (this->senhaFuncionario == senhain){
         cout << "Ola funcionário!" << endl;
            while (atendimento)
            {
                int opc;
                cout << "Qual operacao deseja fazer? (1 - Adicionar conta, 2 - Remover conta, 3 - Sair): ";
                cin >> opc;
                switch (opc)
                {
                    case 1:
                    {
                        int senha, numero;
                        std::string titular, tipo;
                        double saldo;
                        cout << "Digite a senha da conta a ser criada: ";
                        cin >> senha;
                        cout << "Digite o número da conta: ";
                        cin >> numero;
                        cout << "Digite o nome do titular: ";
                        cin >> titular;
                        cout << "Digite o tipo da conta (poupança/corrente): ";
                        cin >> tipo;
                        cout << "Digite a quantidade de dinheiro a ser depositada inicialmente na conta: ";
                        cin >> saldo;
                        adicionaConta(senha, numero, titular, tipo, saldo);
                        break;
                    }
                    case 2:
                        int numero;
                        cout << "Digite o número da conta: ";
                        cin >> numero;
                        removeConta(numero);
                        break;
                    case 3:
                        atendimento = false;
                        break;
                }
            }
    }
    else{
        cout << "Senha invalida" << endl;
    }
}

void Banco::adicionaConta(int senha, int numero, std::string titular, std::string tipo, double saldo){
    Conta *novaConta = new Conta(senha, numero, titular, tipo, saldo);
    Conta *contas = new Conta[++top+1];
    for (int i=0; i<top; i++){
        contas[i] = this->contas[i];
    }
    contas[top] = *novaConta;
    delete [] this->contas;
    this->contas = new Conta[top+1];
    this->contas = contas;
    cout << "Conta adicionada." << endl;
}

bool Banco::removeConta(int numero){
    Conta *contaCliente;
    contaCliente = this->buscaConta(numero);
    if (top == -1){
        cout << "Não existem contas" << endl;
    }
    else if (contaCliente == nullptr)
    {
        cout << "Conta invalida" << endl;
    }
    else
    {
        Conta *contas = new Conta[--top + 1];
        int i=0, j=0;
        while (i<=top){
            if (&this->contas[j] != contaCliente){
                contas[i] = this->contas[j];
                i++;
            }
            j++;
        }
        this->contas = contas;
        cout << "Conta removida, restam "<< top+1 << "." << endl;
        return true;
    }
    return false;
}

File: aula11/banco.py
from contas import Conta

class Banco:
    def __init__(self):
        self.__senhaFunc = 123456
        self.__contas = []
        self.__contas.append(Conta(1, "Joao", 1234, 300))
        self.__contas.append(Conta(2, "Jose", 4567, 800))
        self.__contas.append(Conta(3, "Maria", 7890, 1000))
        self.__contas.append(Conta(4, "Madalena", 8956, 2000))
    
    def buscaConta(self, numero):
        for conta in  self.__contas:
            if numero == conta.numero:
                return conta
        return -1
    
    def atendimentoCliente(self):
        print("Bem vindo ao sistema de atendimento do banco ao cliente")
        numC = int(input("Digite o numero da sua conta: "))
        conta = self.buscaConta(numC)
        if conta == -1:
            print("Conta inválida")
        else:
            print(f"Olá {conta.titular}")
            senha = int(input("Digite a senha: "))
            atendimento = True
            while (atendimento):
                op = int(input("Qual operação deseja fazer? (1 - Ver Saldo, 2 - Saque, 3 - Deposito, 4 - Sair): "))
                if op == 1:
                    print(f"Saldo: R$ {conta.getSaldo(senha)}")
                elif op == 2:
                    valor = float(input("Digite o valor: "))
                    conta.saque(senha, valor)
                elif op == 3:
                    valor = float(input("Digite o valor: "))
                    conta.deposito(valor)
                elif op == 4:
                    atendimento = False

    def atendimentoFuncionario(self):
        print("Bem vindo ao sistema de atendimento do banco ao funcionário")
        senha = int(input("Digite a senha de funcionário: "))
        if senha == self.__senhaFunc:
            print("Ola funcionário!")
            atendimento = True
            while (atendimento):
                op = int(input("Qual operação deseja fazer? (1 - Mudar senha, 2 - Cadastrar cliente, 3 - Sair): "))
                if op == 1:
                    self.__senha = int(input("Digite a nova senha: "))
                elif op == 2:
                    numero = int(input("Digite o número da conta: "))
                    conta = self.buscaConta(numero)
                    if conta!=-1:
                        print("Já existe conta com esse número")
                        return
                    titular = input("Digite o nome do titular: ")
                    senha = int(input("Digite a senha da conta: "))
                    saldo = int(input("Digite a quantidade de dinheiro a ser depositada inicialmente na conta: "))
                    self.__contas.append((numero, titular, senha, saldo))
                    print("Conta adicionada com sucesso!")
                elif op == 3:
                    atendimento = False
        else:
            print("Senha inválida")